change config dir location

webern opened the pull request:
**Is this a bug fix or adding new feature?**

Fixes #246
Replaces #247

Fixes https://github.com/bottlerocket-os/bottlerocket/issues/1111

**What is this PR about? / Why do we need it?**

``` text
change config dir location

Before this change, the driver fails to work on Bottlerocket because it
tries to write files into the host directory /etc/amazon/efs. On
Bottlerocket that directory is not writable by the container.

We want to move the host mount point to /var/amazon/efs, but we need to
support backward compatibility with hosts that may already have configs
written to /etc/amazon/efs.

To solve this, we mount both host paths and check for the presence of
config files at container runtime. If we find files in the old location,
we symlink to that location, otherwise we symlink to the new location.
```

**What testing is done?**

I built my changes to a container image, pushed to a private ECR repo, and changed yaml files accordingly to use my changes. Throughout, I used a canary application from these instructions https://docs.aws.amazon.com/eks/latest/userguide/efs-csi.html which mounts an EFS volume and logs to it every five seconds.

- First, I deployed the current production version to a cluster that had one AL2 node and one Bottlerocket node.
  - I observed that the deployment failed on Bottlerocket and succeeded on AL2.
  - I ssh'ed to the AL2 box and confirmed that files had been written to `/etc/amazon/efs`.
  - I confirmed that the canary application was writing logs to the EFS mount, but only on AL2, Bottlerocket was broken.
- Next I deployed my updated version of the efs driver.
  - I observed that pods on both hosts were now healthy and writing to their EFS mounts.
  - I ssh'ed to AL2 and found that `/etc/amazon/efs` was still being used and `/var/amazon/efs` was empty.
  - I ssh'ed to Bottlerocket and found that `/etc/amazon/efs` was empty and `/var/amazon/efs/ was being used.
- I rebooted both nodes and found that when the pods resumed, they were still able to write to their EFS mounts.
- I also deployed my changed driver into a clean cluster and observed that both AL2 and Bottlerocket used `/var/amazon/efs` and that mounts were working.

webern commented:
>  webern force-pushed the webern:bottlerocket branch from 78d281c to d31f20d

This push fixes the author and a couple of comments. I'm still figuring out the paperwork for the CLA.

webern commented:
>  webern force-pushed the webern:bottlerocket branch from d31f20d to dcee1e2

Hopefully this push will get me passed the CLA?

wongma7 commented:
/check-cla
/ok-to-test

webern commented:
I signed it

webern commented:
/check-cla

webern commented:
I signed it

webern commented:
/check-cla

webern commented:
/retest

webern commented:
Hmmm, OK, I may need some help understanding why the e2e tests are failing. Or maybe a walkthrough of what's actually happening. e.g. does a container with my Go changes get built and used in the test?

webern commented:
> webern force-pushed the webern:bottlerocket branch from dcee1e2 to 7fc9f51

This push brings the helm chart inline with the other podspec.

> webern force-pushed the webern:bottlerocket branch from 7fc9f51 to e7cc126

This push raises code coverage to 100% for the new function.

webern commented:
/retest

wongma7 commented on helm/templates/daemonset.yaml:
```diff
@@ -67,7 +67,9 @@ spec:
               mountPath: /var/run/efs
             - name: efs-utils-config
-              mountPath: /etc/amazon/efs
+              mountPath: /var/amazon/efs
```
I am not 100% sure but I think changes to the helm chart have to be made at the same time as or after the release containing these code changes, i.e. 1.1.0/1.0.1. Otherwise the current helmchart 0.3.0 with driver version 1.0.0 won't work anymore.

LEaving this comment so i can check and be sure...

webern replied:
This change was needed to get CI to pass. I don't understand all the packaging implications, but if we back it out we should expect to CI fail. It was this push that got CI to pass https://github.com/kubernetes-sigs/aws-efs-csi-driver/compare/dcee1e279142b1cdcdc452ab2d688f7077788c19..7fc9f513ab23e9a94470f0311575d9f8011e3195

wongma7 replied:
O yeah, makes sense, thanks let's keep it

wongma7 replied:
You are right, with the new install instruction only after the index gets updated would chart be consumable

wongma7 commented on pkg/driver/config_dir_test.go:
```diff
@@ -0,0 +1,304 @@
+	}
+}
+
+// TestInitConfigDirPreferredError asserts that a symlink an error is returned when a symlink cannot be created to the
```
asserts that ~~a symlink~~ an error is returned...

wongma7 commented on pkg/driver/config_dir_test.go:
```diff
@@ -0,0 +1,304 @@
+	defer os.RemoveAll(dir)
+
+	// create an empty legacy dir
+	legacyDir := filepath.Join(dir, "legacy")
```
since most test cases need to create both a legacyDir and preferredDir (mimicking that kubernetes/kubelet would have created them before the Pod started since they are DirectoryOrCreate HostPaths), could be useful; to refactor it into a function

wongma7 commented on pkg/driver/config_dir_test.go:
```diff
@@ -0,0 +1,304 @@
+	}
+
+	// function under test
+	if err := InitConfigDir(legacyDir, preferredDir, etcAmazonEfs); err != nil {
```
What about a test case vEery similar to this where InitConfigDir is called twice? Arguably redundant, but in practice that is what will happen

wongma7 commented:
/approve
lgtm

thank you very much I really appreciate the numerous clear test cases , I will let @jqmichael  review also before dropping lgtm

wongma7 commented on pkg/driver/config_dir.go:
```diff
@@ -0,0 +1,91 @@
+
+	// if there is already a symlink in place, we have nothing to do
+	if info, err := os.Lstat(etcAmazonEfs); err == nil {
+		if info.Mode()&os.ModeSymlink != 0 {
```
the verify test passed so seems so

webern replied:
Surprisingly `go fmt` removes the spaces that I so purposefully put there in the first place!

webern commented on Dockerfile:
```diff
@@ -39,7 +39,7 @@ RUN yum install amazon-efs-utils-1.26-3.amzn2.noarch -y
 # Otherwise creating a host path volume for that directory will clean up everything inside at the first time.
 # Those static files need to be copied back to the config directory when the driver starts up.
-RUN cp -r /etc/amazon/efs /etc/amazon/efs-static-files
+RUN mv /etc/amazon/efs /etc/amazon/efs-static-files
```
Right, previously you were mounting a host path 'over' this. Now I am writing a symlink in this location, which is simpler if nothing is there already.

jqmichael commented:
/approve

webern commented:
/retest

webern commented:
/retest

wongma7 commented:
/lgtm

wongma7 commented on deploy/kubernetes/base/node.yaml:
```diff
@@ -45,7 +45,9 @@ spec:
-              mountPath: /etc/amazon/efs
+              mountPath: /var/amazon/efs
+            - name: efs-utils-config-legacy
+              mountPath: /etc/amazon/efs-legacy
```
the tests that are failing are all related to the dynamic PR merged just now https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/274.

Could you make the same changes in node.yaml to controller.yaml? shoudl fix it.

webern replied:
Thanks I didn't understand what was going on.

wongma7 replied:
Sorry, what I suggested isn't right either, the controller yaml makes no reference to any config dir

Let me take a closer look

wongma7 commented:
Ok, got it, since the controller makes no reference to any config directories, the new function fails like this:
```
I0106 05:34:47.978762       1 config_dir.go:72] Creating symlink from '/etc/amazon/efs' to '/var/amazon/efs'
F0106 05:34:47.978839       1 main.go:60] config directory '/var/amazon/efs' does not exist: 'stat /var/amazon/efs: no such file or directory'
```
I guess it should be oky to remove the check and just create preferredDir /var/amazon/efs if it doesn't exist? In the node setting we can assume that it will be created by hostpath DirectoryOrCreate, in controller setting we don't care about persisting configs at all.

@kbasv since the controller should only briefly mount efs volumes to delete them is it necessary for its config and such to be persistent? I am saying here no, but could you please confirm.

kbasv commented:
@wongma7 Controller doesn't require persistent configs

webern commented:
> I guess it should be oky to remove the check and just create preferredDir /var/amazon/efs if it doesn't exist? In the node setting we can assume that it will be created by hostpath DirectoryOrCreate, in controller setting we don't care about persisting configs at all.

OK, If I understand correctly, we also need to run in an environment where the mounts that I was expecting are not there. In that case we probably want the contents of `/etc/amazon/efs-static-files` to be available at `/etc/amazon/efs`. I think the simplest thing to do would be to create a symlink `/etc/amazon/efs` -> `/etc/amazon/efs-static-files` when neither of the mounted directories are present.

wongma7 commented:
@webern yes that makes sense to me

webern commented:
This has most of my changes. A simplification of the config dir function to account for runtimes where legacy and preferred host paths are not mounted:

> webern force-pushed the webern:bottlerocket branch from 2e6f31a to 216e3f8 9 minutes ago

I missed a couple of refactor opportunities in the tests and added them in this push:

> webern force-pushed the webern:bottlerocket branch from 216e3f8 to ec2d8df 2 minutes ago

I rebased and fixed conflicts in main.go:

> webern force-pushed the webern:bottlerocket branch from ec2d8df to 5ccb581 2 minutes ago

webern commented:
> webern force-pushed the webern:bottlerocket branch from 5ccb581 to 45acb17 20 seconds ago

I can't make os.CreateDirAll fail reliably in any environment, so I deleted that test. Unfortunately I can't give you 100% coverage as that error return cannot reasonably be exercised.

webern commented:
> @webern yes that makes sense to me

Actually the right solution was to create an empty dir since the configs will be copied into it.

wongma7 commented on pkg/driver/config_dir_test.go:
```diff
@@ -0,0 +1,243 @@
+	defer cleanup(t, dir)
+
+	// create legacy dir and a fake pre-existing conf file
+	// create an empty legacy dir
```
nit extra comment here

wongma7 commented:
/lgtm

beautiful, thank you

Next steps for me:
- Test the upgrade scenario end-to-end w/ both helm and kustomize. (cc https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/213)
- Make a helm chart and release of the driver. Probably will do minor version bump.

webern commented:
I fixed the incorrect comment. We needed a retest anyway. The 12 failed tests looked like timeouts, which is what I remember failed tests looking like in the past when a retest passed. Unfortunately fixing that comment removed the lgtm flag.

wongma7 commented:
/lgtm

Yes, just checked logs too and this time the containers all became ready so I don't think it has anything to do with this PR. will repro locally tomorrow if happens again
